Improve docstrings for `filesize` and `stat`

KronosTheLate opened the pull request:
The docstring for `filesize` failes to mention how paths are joined, and document the full-path-as-single-argument method. I know that this version of the docstring might be a regression when it comes to accurately reflecting the underlying methods, but it does serve the propose of documenting how it should be called much better, which is IMO the purpose of docstrings.

I also updated the docstring for `stat` to drop the shortcut, and be more explicit about the intended format for arguments to be slurped and stating that `joinpath` is used.

KronosTheLate committed:
Document internal calls to joinpath

KronosTheLate committed:
Document internal calls to joinpath

KronosTheLate committed:
Merge pull request #1 from JuliaLang/master

merger actual master into mine

KronosTheLate committed:
Merge pull request #2 from KronosTheLate/master

update feature branch

nsajko commented on base/stat.jl:
```diff
@@ -201,9 +201,11 @@ end
-    stat(joinpath...)
+    stat(path_elements...)
+
+Return a structure whose fields contain information about the file. 
```
```suggestion
Return a structure whose fields contain information about the file.
```

LilithHafner committed:
Fix whitespace

Co-authored-by: Neven Sajko <s@purelymail.com>

LilithHafner committed:
Fix whitespace

Co-authored-by: Neven Sajko <s@purelymail.com>

LilithHafner commented:
We should probably keep `filesize` in sync with all of

`ispath`, `isfifo`, `ischardev`, `isdir`, `isblockdev`, `isfile`, `issocket`, `issetuid`, `issetgid`, `issticky`, `uperm`, `gperm`, `operm`, `filemode`, `filesize`, `mtime`, and `ctime`.

They are all defined here:

https://github.com/JuliaLang/julia/blob/6c70bf784999b478bc176fe594738f9746a1dcfd/base/stat.jl#L487-L507

KronosTheLate committed:
Merge branch 'improve_docstring_filesize' of https://github.com/KronosTheLate/julia into improve_docstring_filesize

KronosTheLate committed:
Revamp docstrings in stat.jl

KronosTheLate committed:
Refine docstring for StatStruct

KronosTheLate committed:
fix ordering of signatures

KronosTheLate commented:
> We should probably keep filesize in sync with all of ...

That turned out to be a bigger job than I expected. I ended up putting in the work to clean up the docstrings throughout the module. Here is a highlight of the changes made, and relevant desicions:
- I changed `file` to `path` in all docstrings where the intended usage is passing a path
- I added signatures for `stat_struct`, which replaces the signatures that take in `file`, as it is more explicit on how to pass in a "file". This can confuse newer users, but felt important to accurately and completely document potential usage of the function. This signature comes last to minimize that potential confusion.
  - I did _not_ add signatures for `stat_struct` for functions that are clearly intended to operate on paths rather than file objects/descriptors, namely the functions `isfile`, `ispath`, `isdir`, `issocket`, and `islink`. I figured it would only add confusion, and that for these functions, the method that take `StatStruct`s as inputs are best considered an implementation detail rather than valid usage of the function.
- I cleaned up the language in terms of accuracy, such as changing "Return `true` if `path` is a symbolic link" into "Return `true` if `path` points to a symbolic link". The original version was plain wrong when taken literally, as `path` is a string and not a symbolic link. This makes the language a bit more verbose, but makes it so that when you read each word carefully you get the right conclusion, which was impossible before without also making the right assumptions.
- The original PR mentions the internal usage of `joinpath` in the docstring. Adding this to each and every docstring felt like too much, and the `functionname(path_elements...)` signature shoud document this sufficiently. I have kept the explicit statement "If multiple arguments are given, they are joined by [`joinpath`](@ref)" in the docstring for `stat`, so that if a user keeps drilling down, they will reach the `stat` function and can read it there. Adding the `stat_struct` signatures, along with the new and improved docstring for `StatStruct`, makes it more possible to understand that `stat` is used internally in all cases, which is why it felt adequate to mention `joinpath` explicitly only once, and in the docstring for `stat`
- At least one punctuation, and one plural-replated gramatical error.

KronosTheLate committed:
fix trailing whitespace

LilithHafner commented on base/stat.jl:
```diff
@@ -417,68 +460,107 @@ isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
 
-Like [`uperm`](@ref) but gets the permissions for people who neither own the file nor are a member of
-the group owning the file
+Like [`uperm`](@ref) but gets the permissions for people who neither own the 
```
```suggestion
Like [`uperm`](@ref) but gets the permissions for people who neither own the
```

LilithHafner commented on base/stat.jl:
```diff
@@ -417,68 +460,107 @@ isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
+
+julia> has_read_permission(path) = bitstring(uperm(path))[end-2]=='1';
+
+julia> has_read_premission("dummy_file")
```
```suggestion
julia> has_read_permission("dummy_file")
```

LilithHafner commented on base/stat.jl:
```diff
@@ -417,68 +460,107 @@ isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
 | 04    | Read Permission    |
 
-For allowed arguments, see [`stat`](@ref).
+The fact that a bitfield is returned means that if the premission
```
```suggestion
The fact that a bitfield is returned means that if the permission
```

LilithHafner commented on base/stat.jl:
```diff
@@ -308,33 +323,50 @@ function filemode_string(mode)
+    ctime(path_elements...)
+    ctime(stat_struct)
+
+Return the unix timestamp of when the metadata of the file at `path` was last modified, 
```
```suggestion
Return the unix timestamp of when the metadata of the file at `path` was last modified,
```

LilithHafner commented on base/stat.jl:
```diff
@@ -417,68 +460,107 @@ isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
+julia> bitstring(ans)
+"00000110"
+
+julia> has_read_permission(path) = bitstring(uperm(path))[end-2]=='1';
```
```suggestion
julia> has_read_permission(path) = uperm(path) & 4 != 0;
```
I think it's better to model bit-twiddling without using `bitstring`; while performance of `bitstring` is negligible compared to the stat call, it's still IMO good form to avoid stringification.

KronosTheLate replied:
While I believe you make a technically good point, your change killed the pedagogical aspect for me completely, which was my whole point in adding it. I am unable to parse `uperm(path) & 4 != 0` in my head, but indexing into the bitstring, which is printed above, is trivial, and makes very clear what is going on.

Peraps it is possible to add your version as a comment? Along the lines of
```julia
has_read_permission(path) = bitstring(uperm(path))[end-2]=='1';    # More efficient, less readable:  uperm(path) & 4 != 0
```

LilithHafner replied:
Would `uperm(path) & 0b00000100 != 0` be easier to parse?

KronosTheLate replied:
Yhea, it would. I'll commit that.

LilithHafner commented on base/stat.jl:
```diff
@@ -351,22 +383,29 @@ ispath(path::AbstractString) = ispath(String(path))
+    isfifo(stat_struct) -> Bool
 
-Return `true` if `path` is a FIFO, `false` otherwise.
+Return `true` if the file at `path` is FIFO, `false` otherwise.
+Or the equivalent as indicated by the file descriptor `stat_struct`.
```
```suggestion
Return `true` if the file at `path` or file descriptor `stat_struct` is FIFO, `false` otherwise.
```
I think this is more clear, do you agree? I am happy to use your language if you prefer.

KronosTheLate replied:
No that looks good to me.

LilithHafner replied:
Would you please make this change in all the other places where it applies. Any docstring where
- There's a description for path and then an "or equivalent" clause or similar and
- It would be more clear to use the "file at `path` or file descriptor `stat_struct`" language

KronosTheLate replied:
Done. It was a bit more ankways in the "points at" cases, but I think I found a good compromise.

LilithHafner reviewed (commented):
Thank you! This is clearly a significant improvement and I appreciate the work you put into it. I agree with all your changes and have a few minor tweaks I'd recommend.

KronosTheLate committed:
Remove trailing whitespace

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

KronosTheLate committed:
Fix typo

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

KronosTheLate committed:
Fix typo again

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

KronosTheLate committed:
Another trailing whitespace

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

KronosTheLate committed:
Change phrasing

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

KronosTheLate committed:
More efficient has_read_permission

KronosTheLate committed:
merge master into branch

KronosTheLate committed:
Fix ankward wording

LilithHafner commented on base/stat.jl:
````diff
@@ -417,68 +457,108 @@ isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
+See also [`gperm`](@ref) and [`operm`](@ref).
+
+```jldoctest
+julia> touch("dummy_file")  # Create test-file without contents;
````
```suggestion
julia> touch("dummy_file");  # Create test-file without contents;
```
To fix the doctest

KronosTheLate replied:
Haha whoops. I all manually change this, and also remove the semicolon from the comment

KronosTheLate committed:
Fix doctest

LilithHafner commented:
Lovely, thanks so much!
